### DISCUSSION TITLE
Additional node modules for extensions

### DISCUSSION THREAD

**Discussion by Feuerhamster** [directus/directus#5052] [Archive]
I would like to create custom API endpoints and hooks, but I need additional node modules in these.
Two questions about this:
Which node modules are all available there already?
How do I install additional node modules to it?

My Directus instance is installed via npx, so it is a production system.

**Comment by rijkvanzanten** [comment]
> Which node modules are all available there already?

This depends on the location of your extensions folder. If it's next to the package.json of your project, you can theoretically use any of the dependencies that exist under the Directus package (which are in `node_modules`). That being said, there's no guarantee that those dependencies will always be there for every version, so I'd recommend installing them in your project regardless.

> How do I install additional node modules to it?

The same as any other "regular" node project! An `npm install xxx` + a `const xx = require('xx');` will do the trick

**Reply by benhaynes** [reply]
Think we should add a _Tip_ in the docs regarding this?

**Reply by rijkvanzanten** [reply]
To me this is a little too close to _how to use node_ compared to the docs on what is Directus specific, but it could be a good addition for a more in-depth example (cc @nickrum )

**Reply by nickrum** [reply]
While I agree that it is close to _how to use node_, I can see where the confusion is coming from. The docs for endpoints and hooks only really mention a single js file. Without knowing how directus actually uses those extensions, it might not be clear that those single js files could actually depend on other files or packages.
Regarding more in-depth examples, I think it would be nice if we had a few "official" extensions that also serve as practical, in-depth examples.

**Comment by arladmin** [comment]
Hi @rijkvanzanten , i have the same question, but for the docker based setup.

How to import external node modules (e.g. dayjs etc.) for use in custom endpoints/hooks ?

Going by the docs, i was only able to implement a simple js file with the min. code needed for the particular extension. But how to import from npm, because ```const dayjs = require('dayjs')``` doesn't work?

**Reply by nickrum** [reply]
You could either bundle the dependency into your extension or use the Directus Docker image as a base for a custom image where you install the necessary dependency.

**Reply by arladmin** [reply]
@nickrum Thanks, will try this out!
